src/compiler/checker.ts

```diff
@@ -336,6 +341,13 @@ namespace ts {
                         target[id] = source[id];
                     }
                     else {
+                        if (target === globals && source !== builtinGlobals) {

```

Comment:
We can get some wonky undefined behavior while typechecking if we actually let them merge (eg, this is a class which is also `undefined`), so we have to guard against it in the merge function so we don't let it poison the type checking process. Outright overriding the symbol is _okay_ and is what #5387 allowed, but merging with the builtin undefined symbol is unallowable.